lib/render/SliceRenderer.cpp

```diff
@@ -178,144 +188,231 @@ int SliceRenderer::_resetBoxCache()
+    // Perform convex hull on our list of 2D points,
+    // which defines the outer edges of our polygon
+    stack<glm::vec2> orderedTwoDPoints = convexHull(unorderedTwoDPoints, sizeof(unorderedTwoDPoints) / sizeof(unorderedTwoDPoints[0]));
+    return orderedTwoDPoints;
+}
 
-            dataValues[index] = varValue;
+std::vector<glm::vec2> SliceRenderer::_makeRectangle2D(const std::vector<_vertex> &vertices, stack<glm::vec2> &orderedTwoDPoints) const
+{
+    std::vector<glm::vec2> rectangle2D = {glm::vec2(), glm::vec2()};
+    while (!orderedTwoDPoints.empty()) {
+        glm::vec2 vertex = orderedTwoDPoints.top();
+        if (vertex.x < rectangle2D[0].x) rectangle2D[0].x = vertex.x;
+        if (vertex.y < rectangle2D[0].y) rectangle2D[0].y = vertex.y;
+        if (vertex.x > rectangle2D[1].x) rectangle2D[1].x = vertex.x;
+        if (vertex.y > rectangle2D[1].y) rectangle2D[1].y = vertex.y;
+        orderedTwoDPoints.pop();
+    }
+    return rectangle2D;
+}
 
-            index += 2;
-            coords[X] += deltas[X];
+// Map our rectangle's 2D edges back into 3D space, to get an
+// ordered list of vertices for our data-enclosing polygon.
+std::vector<glm::vec3> SliceRenderer::_makePolygon3D(const std::vector<_vertex> &vertices, stack<glm::vec2> &polygon2D) const
+{
+    std::vector<glm::vec3> polygon3D;
+    while (!polygon2D.empty()) {
+        glm::vec2 twoDPoint = polygon2D.top();
+        for (auto &vertex : vertices) {
+            if (twoDPoint.x == vertex.twoD.x && twoDPoint.y == vertex.twoD.y) {
+                polygon3D.push_back(glm::vec3(vertex.threeD.x, vertex.threeD.y, vertex.threeD.z));
+                polygon2D.pop();
+                break;
+            }
         }
-        coords[Y] += deltas[Y];
     }
+    return polygon3D;
 }
 
-void SliceRenderer::_populateDataXZ(float *dataValues, Grid *grid) const
+std::vector<glm::vec3> SliceRenderer::_makeRectangle3D(const std::vector<_vertex> &vertices, stack<glm::vec2> &polygon2D) const
 {
-    std::vector<double> deltas = _calculateDeltas();
-    float               varValue, missingValue;
-    std::array<double, 3> coords = {0.0, 0.0, 0.0};
-    coords[X] = _cacheParams.domainMin[X];
-    coords[Y] = _cacheParams.boxMin[Y];
-    coords[Z] = _cacheParams.domainMin[Z];
-
-    int index = 0;
-    for (int j = 0; j < _textureSideSize; j++) {
-        coords[X] = _cacheParams.domainMin[X];
-
-        for (int i = 0; i < _textureSideSize; i++) {
-            varValue = grid->GetValue(coords);
-            missingValue = grid->GetMissingValue();
-            if (varValue == missingValue)
-                dataValues[index + 1] = 1.f;
-            else
-                dataValues[index + 1] = 0.f;
-
-            dataValues[index] = varValue;
+    // Define a rectangle that encloses our polygon in 3D space.  We will sample along this
+    // rectangle to generate our 2D texture.
+    std::vector<glm::vec3> rectangle3D = {glm::vec3(), glm::vec3(), glm::vec3(), glm::vec3()};
```

Comment:
Is there a reason why inverseProjection is a lamba and not a function defined at the file scope? You redefine the same lambda later in the code, which can lead to a maintenance problem. What if one gets changed and the other doesn't?

Reply:
That's a good point.  Fixed.